File: 프로그래머스/2/72412. 순위 검색/순위 검색.py
from itertools import combinations
from collections import defaultdict
import bisect

def solution(info, query):
    answer = []

    def combi(arr, select_num): return list(combinations(arr, select_num))

    info_score = defaultdict(list)
    info_score[""] = []

    for v in info:
        val = v.split(" ")
        score = int(val.pop())
        info_score[""].append(score)

        for i in range(1, 5):
            comb = combi("0123", i)
            for c in comb:
                key = ''.join([val[int(index)] for index in c])
                info_score[key].append(score)

    for key in info_score: info_score[key].sort()

    for v in query:
        v = v.replace(" and ", "").replace("-", "").split(" ")
        key = v[0]
        score = int(v[1])

        if key in info_score:
            scores = info_score[key]
            index = bisect.bisect_left(scores, score)
            answer.append(len(scores) - index)
        else: answer.append(0)

    return answer

File: 프로그래머스/2/72412. 순위 검색/순위 검색.js
function binarySearch(arr, target) {
    let left = 0, right = arr.length - 1;

    while (left <= right) {
        let mid = Math.floor((left + right) / 2);
        if (arr[mid] >= target) right = mid - 1;
        else left = mid + 1;
    }

    return left;
}

function getInfos(info) {
    const infos = {};
    info.forEach(infoStr => {
        const arr = infoStr.split(" ");
        const score = parseInt(arr.pop());
        const key = arr.join(" ");
        if (infos[key]) infos[key].push(score);
        else infos[key] = [score];
    });

    for (const key in infos) infos[key].sort((a, b) => a - b);

    return infos;
}

function getResult (infos, query, score) {
  const infosKey = Object.keys(infos);
  return infosKey
    .filter(key => query.every(v => key.includes(v)))
    .reduce((a, key) => a + infos[key].length - binarySearch(infos[key], score), 0);
}

function solution(info, query) {
  let answer = [];
  const infos = getInfos(info)
  query
    .map(q => q.split(/ and | |-/i).filter(v => v !== ""))
    .forEach(q => {
      const score = q.pop();
      const result = getResult(infos, q, score);
      answer.push(result)
    })
    
  return answer;
}